src/python/pants/bin/daemon_pants_runner.py

```diff
@@ -36,20 +36,28 @@ class DaemonExiter(Exiter):
 
   @classmethod
   @contextmanager
-  def override_global_exiter(cls, maybe_shutdown_socket: MaybeShutdownSocket, finalizer: Callable[[], None]) -> None:
-    with ExceptionSink.exiter_as(lambda previous_exiter: cls(maybe_shutdown_socket, finalizer, previous_exiter)):
+  def override_global_exiter(
+    cls, maybe_shutdown_socket: MaybeShutdownSocket, finalizer: Callable[[], None],
+  ) -> Iterator[None]:
+    with ExceptionSink.exiter_as(
+      lambda previous_exiter: cls(maybe_shutdown_socket, finalizer, previous_exiter)
+    ):
       yield
 
-  def __init__(self,
+  def __init__(
+    self,
     maybe_shutdown_socket: MaybeShutdownSocket,
     finalizer: Callable[[], None],
-    previous_exiter: Exiter):
-
-    super().__init__(exiter=previous_exiter)
+    previous_exiter: Optional[Exiter],
+  ) -> None:
+    if previous_exiter:
+      super().__init__(exiter=previous_exiter._exit)
+    else:
+      super().__init__()
```

Comment:
Sanity check would be helpful here. This is an actual change to exception sink handling for pantsd. I don't think the original was working how we expected. 

The constructor for `Exiter` expects a function that gets set to `self._exit`. This is then used by `Exiter.exit()`.  It defaults to `sys.exit`. 

The old `daemon_pants_runner.py` code here did not account for two things:
1) `previous_exiter` may be `None`. `ExceptionSink` defaults to `None` for the class property. In practice, it's always defined because we set the value to `sys.exit` when importing `exception_sink.py`, but ideally this constructor will not rely on that assumption.
2) we were passing an `Exiter` instance to `Exiter.__init__()`, rather than a function. So, `self._exit` was being set to an `Exiter`. I think this worked because we define `Exiter.__call__()`, but does not seem intended. Instead, we should pass to the constructor the underlying `_exit` property from `previous_exiter`.

cc @cosmicexplorer @illicitonion @blorente

Reply:
> In practice, it's always defined...but ideally this constructor will not rely on that assumption.

This is great! Could we add a comment to that effect here?

> I think this worked because we define `Exiter.__call__()`, but does not seem intended.

Hm, my impression was that this *was* intended (this was my interpretation of the "sys.exit fungibility" comment in `exiter.py`), and relying on the public `__call__()` seems like it might be preferable to the private `._exit` property?

Reply:
> I think I would have accepted the previous approach of using the default Exiter(exiter=sys.exit) as well -- I really wasn't sure how to structure that code at first and it seems like it might be easier to avoid having to add extra handling here for the None case?

It caused a test to fail because we printed `ParserError: ` instead of `Exception message: ` :/ That's why I reverted the change.

--

> Hm, my impression was that this was intended (this was my interpretation of the "sys.exit fungibility" comment in exiter.py)

I'm not 100% confident on this, but that "feels" wrong. In particular, my concern is that setting `super().__init__(exiter=previous_exiter)` means that `self._exiter` will now be equivalent to `self.exit()`, per the implementation of `__call__()`:

https://github.com/pantsbuild/pants/blob/eed447a2971a417faca9b5744c8f38b7387e8604/src/python/pants/base/exiter.py#L41-L43

When someone calls `DaemonExiter.exit()`, then `exit()` will run normally until the last line below, where it will recursively call itself? Possibly missing something, but that's how I read it: 

https://github.com/pantsbuild/pants/blob/eed447a2971a417faca9b5744c8f38b7387e8604/src/python/pants/base/exiter.py#L45-L78

Also, note the docstring of the constructor. According to that, I don't think we intended to allow passing an `Exiter` instance to the constructor for `Exiter`.

https://github.com/pantsbuild/pants/blob/eed447a2971a417faca9b5744c8f38b7387e8604/src/python/pants/base/exiter.py#L30-L33

Reply:
Going to merge this PR now to unblock the followup MyPy PRs. If we realize this particular `daemon_pants_runner.py` change was a mistake, it will be trivial to revert that portion of this PR.